### DISCUSSION TITLE
Questions on how to deploy a custom client

### DISCUSSION THREAD

**Discussion by ocularvr** [mozilla/hubs#3060] [Hubs Cloud Questions] [answered]
I was wondering if anyone could guide me on how to 
1. Modify the client
2. Deploy the client
This document is of no use to me and others in the forum:
https://hubs.mozilla.com/docs/hubs-cloud-custom-clients.html
I get to the bit:

_Now that you have the Hubs repository cloned you'll need to install some dependencies. You'll need Node JS installed first. Then you'll install the hubs dependencies. We recommend using npm ci instead of npm install so that you always use the versions of modules in the package-lock.json file._

`npm ci
`

I don't get past this.

**Comment by robinkwilson** [comment, marked as answer]
Have you downloaded Nodejs? https://www.npmjs.com/get-npm
Open your terminal or command prompt and run the `npm -v` or `node -v`. 

Once you have the Hubs repo on your machine navigate to the directory with your just cloned Hubs repo (using `ls` (Mac or linux) or `dir` (Windows) to check current directory and change directory with `cd` (both)) , then run `npm ci` in your terminal/command prompt. 

Hope this helps

**Reply by ocularvr** [reply]
Thanks for your feedback, just one question:

https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-up-node-on-ec2-instance.html
Is this the right path to take before running

`npm -v`

?

**Reply by robinkwilson** [reply]
Oh, this should be on your computer, not on your EC2 server.

`npm run login` and `npm run deploy` handles deploying the custom client to your Hubs Cloud stack. 

I haven't watched this in its entirety, but here's a video on using the command line, github, forking a repo, cloning a repo: https://www.youtube.com/watch?v=OODDLyvePr8


(btw `npm -v` only gets the version of `npm` you're running)

**Reply by ocularvr** [reply]
Thank for your feedback.
Just watched the video 5 times and still at a total loss. I wont give up!
I re-read the document https://hubs.mozilla.com/docs/hubs-cloud-custom-clients.html. 

My eyes started tearing after reading the word "Branch" three times in the same sentence:

"When working on a custom client, you should base your code on the hubs-cloud branch branch, which will be the branch that has client changes compatible with Hubs Cloud servers. When server upgrades occur (approximately once a month), you should merge with the latest changes to this branch. Changes due to upgrades will be visible in merged PRs on the branch."

I really have no idea what this means. Will I make changes and have to wait one month to see those changes?

Oh please, I really need a document that says how someone with 5 years of experience deploying LAMP stuff could modify a hubs client and deploy a hubs client on aws. I swear I wont post anything for a month if you guys do this two page document.

Im really trying here, even with Node.js on my local windows 10, you get a powershell installed and then you need to install git on top of that. Come on, I know im dumb but the assumptions on Mozillas part are just way toooooo much. Every option (github/node.js/aws/powershell) is a labyrinth of settings and to assume that this is just a detail is wrong. 

My head just popped!

<img src=https://www.sfweekly.com/wp-content/uploads/2016/02/gilsheadexplodes.gif>

**Reply by robinkwilson** [reply]
No worries! If you know any developers that have worked with github before, they can help you :) No need to do anything fancy in aws, you can use the command line and the package.json commands like `npm run deploy`.
